ops/hookcmds/_relation.py

```diff
@@ -0,0 +1,150 @@
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+
+    Args:
+        app: Get the relation data for the overall application, not just a unit
+        id: The ID of the relation to get data for, or ``None`` to get data for
+            the relation that triggered the current hook.
+        key: The specific key to get data for, or ``None`` to get all data.
+        unit: The unit to get data for, or ``None`` to get data for all units.
+    """
+    args = ['relation-get', '--format=json']
+    if id is not None:
+        args.extend(['-r', str(id)])
+    if app:
+        args.append('--app')
+    if key is not None:
+        args.append(key)
+    if unit is not None:
+        args.append(unit)
+    result = json.loads(run(*args))
+    if key is not None:
+        return cast('str', result)
+    return cast('dict[str, str]', result)
+
+
+def relation_ids(name: str) -> list[str]:
+    """List all relation IDs for the given endpoint.
+
+    Args:
+        name: the endpoint name.
+    """
+    return cast('list[str]', json.loads(run('relation-ids', name, '--format=json')))
+
+
+def relation_list(id: int | None = None, *, app: bool = False) -> list[str]:
+    """List relation units.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+
+    Args:
+        id: The ID of the relation to list units for, or ``None`` to list units
+            for the relation that triggered the current hook.
+        app: List remote application instead of participating units.
+    """
+    args = ['relation-list', '--format=json']
+    if app:
+        args.append('--app')
+    if id is not None:
+        args.extend(['-r', str(id)])
+    return cast('list[str]', json.loads(run(*args)))
+
+
+def relation_model_get(id: int | None = None) -> RelationModel:
+    """Get details about the model hosting a related application.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+
+    Args:
+        id: The ID of the relation to get data for, or ``None`` to get data for
+            the relation that triggered the current hook.
+    """
+    args = ['relation-model-get', '--format=json']
+    if id is not None:
+        args.extend(['-r', str(id)])
+    return RelationModel._from_dict(cast('RelationModelDict', json.loads(run(*args))))
+
+
+def relation_set(
+    data: Mapping[str, str],
+    id: int | None = None,
+    *,
+    app: bool = False,
+):
+    """Set relation settings.
+
+    Note that ``id`` can only be ``None`` if the current hook is a relation
+    event, in which case Juju will use the ID of the relation that triggered the
+    event.
+
+    Args:
+        data: The relation data to set.
```

Comment:
to document:
- only updates the supplied keys
- if you want to whack a field, set empty string value

Reply:
Done